### PR TITLE
Fix opensearch theme version

### PR DESCRIPTION
**Pull request by zuochengding**
Signed-off-by: Zuocheng Ding <zding817@gmail.com>

### Description
This PR will update the old kibana theme version and tag to `v1` and `v2`. 

[EDIT] : The new revision will change the default theme version to `v1` and remove the beta theme
 
### Issues Resolved
 - https://github.com/opensearch-project/OpenSearch-Dashboards/issues/494
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
    - [X] `yarn test:jest`
    - [X] `yarn test:jest_integration`
    - [X] `yarn test:ftr`
- [X] New functionality has been documented.
- [X] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by kavilla** [comment]
Nit: Can we update the commit to include more details and have a description. Doesn't have to be much and you can add the issue in there as well.

**Comment by kavilla** [comment]
For tracking purposes related to this PR. Can we say we manually verified this will not break backwards compatibility.

**Comment by zuochengding** [comment]
> For tracking purposes related to this PR. Can we say we manually verified this will not break backwards compatibility?

Yes, the PR has no logical change but just updates the naming of the themes. This will only affect the UI. From my local testing, there is no behavior change. In theory, there should have any regression issue. However, one thing I have to mention here is that `v2` (or the original `v8(beta)`) has some SCSS errors in console.

**Review by kavilla — DISMISSED** [review]

**Review by ananzh — APPROVED** [review]
LGTM

**Comment by tmarkley** [comment]
I think we should consider just using the default and removing this option as Jules mentioned in #494.

**Comment by jcgraybill** [comment]
Yes, let's get @ahopp to take a look at this before merging - to weigh in on what to call these two alternative themes, or whether to just standardize on a single theme.

**Comment by ahopp** [comment]
@tmarkley sorry for the delay! I'm not seeing a ton of value in the "secondary" theme. For now, I think it would be better to standardize on a single theme. Do we have a good view of the differences between the two themes?

**Comment by zuochengding** [comment]
> @tmarkley sorry for the delay! I'm not seeing a ton of value in the "secondary" theme. For now, I think it would be better to standardize on a single theme. Do we have a good view of the differences between the two themes?

The theme (v8 beta) does't seem fully finished, I have noticed some layout cutting in UI (and errors in console as well) for the second theme (v8 beta). From this standing point, i think it is good to remove.

**Comment by ahopp** [comment]
Roger, in that case it makes sense to align on the "v7" and begin any work on new theming can be resolved at a latter date. Thanks!

**Comment by kavilla** [comment]
If any case, if we remove the option lets make sure it doesn't break any backwards compatibility on start up if someone from the legacy app had this setting to the beta.

**Review by boktorbb-amzn — DISMISSED** [review]
I agree with backwards compatibility verification; otherwise, looks good to me

File: src/core/server/ui_settings/settings/theme.test.ts [hunk 1]
```diff
@@ -58,13 +58,10 @@ describe('theme settings', () => {
     const validate = getValidationFn(themeSettings['theme:version']);
 
     it('should only accept valid values', () => {
```
Comment (tmarkley) [inline src/core/server/ui_settings/settings/theme.test.ts:60]: I think we should still have tests for v7 and v8 to make sure the changes are backwards compatible.

File: packages/osd-ui-shared-deps/theme.ts [hunk 2]
```diff
@@ -38,26 +38,15 @@ export type Theme = typeof LightTheme;
 
 // in the OpenSearch Dashboards app we can rely on this global being defined, but in
 // some cases (like jest) the global is undefined
-export const tag: string = globals.__osdThemeTag__ || 'v7light';
-export const version = tag.startsWith('v7') ? 7 : 8;
+export const tag: string = globals.__osdThemeTag__ || 'v1light';
 export const darkMode = tag.endsWith('dark');
 
-export let euiLightVars: Theme;
-export let euiDarkVars: Theme;
```
Comment (tmarkley) [inline packages/osd-ui-shared-deps/theme.ts:46]: Doesn't EUI still need these variables?

**Comment by tmarkley** [comment]
Should we remove the UI since there is only one theme?

**Comment by kgcreative** [comment]
> Should we remove the UI since there is only one theme?

We shouldn't remove the UI, since it's a valid setting, but we should display this as a disabled read only input field instead of a drop down selector. If we provide alternate themes in the future, we will want to leverage this setting.

**Comment by zuochengding** [comment]
FYI, The latest revision is to add a `readOnly` flag to the theme version UI setting. This will hide the UI from `Advanced setting`. By doing so, we are able to make sure all customers can see the same UI/UX even they are from legacy app with (v8 beta) theme configured. We will use consistent theme `v1` and the code is still there if we need add more theme support in future.

**Comment by tmarkley** [comment]
Oh, I thought we agreed to not remove the UI completely based on what Kevin said above? I think we can keep the UI but make the theme dropdown disabled/read only with the v1 theme selected. We should also mark this as a "breaking" change and target a `2.0.0` release since we'll be removing the v8 theme and defaulting all users to the renamed v1 theme.

**Comment by zuochengding** [comment]
> Oh, I thought we agreed to not remove the UI completely based on what Kevin said above? I think we can keep the UI but make the theme dropdown disabled/read only with the v1 theme selected. We should also mark this as a "breaking" change and target a `2.0.0` release since we'll be removing the v8 theme and defaulting all users to the renamed v1 theme.

Yes, that was my initial direction.  However, Based on the experimental testings yesterday. there are some limitations here:
1. It is not straightforward to create a readonly text over there, the whole css files comes from EUI CSS dependencies, we don't have pure readonly text type builtin for ui_setting items. 
2. Also, I was having challenges to override the cached user value in a general way, a hack string conversion for a particular user value doesn't looks promising.
3. The build-in `readOnly` will actually hide the UI instead making it not editable surprised me as well. It makes some senses in the point that `if it is readonly, then it shouldn't display in a configurable setting config page`
4. This actually solves the biggest regression problem in my opinion, the user won't see a confusing UI even though they are a cached user setting in their previous revision.

**Comment by kgcreative** [comment]
@tmarkley @zuochengding - If the "standard pattern" for advanced settings is to hide read only settings, let's stick to the convention for now. We should revisit whether that is a direction that we want to keep moving forward, but that's out of scope for this particular issue.

**Comment by tmarkley** [comment]
@zuochengding are we targeting `v2.0.0` for this change? Can we add tests that confirm that the `v1` theme is properly set when either `v7` or the `v8` beta theme were selected previously?

**Comment by zuochengding** [comment]
> @zuochengding are we targeting `v2.0.0` for this change? Can we add tests that confirm that the `v1` theme is properly set when either `v7` or the `v8` beta theme were selected previously?

Yes, we are targeting `v2.0.0` and I will add more tests in next revision

File: src/core/server/ui_settings/settings/theme.ts [hunk 3]
```diff
@@ -51,14 +51,19 @@ export const getThemeSettings = (): Record<string, UiSettingsParams> => {
...
+      readonly: true, // This flag will hide the `theme:version` UI from Advancend Settings.
       description: i18n.translate('core.ui_settings.params.themeVersionText', {
         defaultMessage: `Switch between the theme used for the current and next version of OpenSearch Dashboards, A page refresh is required for the setting to be applied.`,
       }),
       requiresPageReload: true,
-      schema: schema.oneOf([schema.literal('v7'), schema.literal('v8 (beta)')]),
+      schema: schema.oneOf([
+        schema.literal('v1'),
+        schema.literal('v7'),
+        schema.literal('v8 (beta)'),
```
Comment (tmarkley) [inline src/core/server/ui_settings/settings/theme.ts:65]: Why can't we remove these from the schema?

File: src/core/server/ui_settings/settings/theme.test.ts [hunk 4]
```diff
@@ -58,12 +58,12 @@ describe('theme settings', () => {
...
-      expect(() => validate('v7')).not.toThrow();
-      expect(() => validate('v8 (beta)')).not.toThrow();
```
Comment (tmarkley) [inline src/core/server/ui_settings/settings/theme.test.ts:62]: Can we keep these and expect them to throw?

**Review by tmarkley — CHANGES_REQUESTED** [review]
I think before we merge this we should probably have integration and functional tests that confirm that no matter what setting was applied before updating, `v1` is always applied.

File: src/core/server/ui_settings/settings/theme.ts (hunk 3 above)
Comment (zuochengding) [reply src/core/server/ui_settings/settings/theme.ts:65]: This is for backward compatibility, user's injectedMetadata from previous version is `readonly` and there is no way to modify it. The current system will do an initial schema validation, so you will see some warnings in console if we remove this fields. The way I proposed is to override the user injectedData after the data loaded (deep copied) into the memory (cache), then I override it as default value `v1`.

File: src/core/server/ui_settings/settings/theme.test.ts (hunk 4 above)
Comment (zuochengding) [reply src/core/server/ui_settings/settings/theme.test.ts:62]: same reason as above, It will validate based on the schema we defined

**Comment by zuochengding** [comment]
> I think before we merge this we should probably have integration and functional tests that confirm that no matter what setting was applied before updating, `v1` is always applied.

I think the unit test might be sufficient because it covered the main concern for naming change. Just to make sure we are on same page, `v1` isn't something new but just a renamed theme of `v7`. Compared with previous, only one theme will be loaded no matter what name it is. The only concern would be what is name will displayed to customer.  The unit test will verify the override . Functional testing is something we might can do but since we have actually disabled/hided the from the `Advanced Setting`, not very sure if the value can be retrieved from UI level verification. I don't see much benefit for this since it is covered in unit test

**Comment by kavilla** [comment]
@zuochengding, @tmarkley, in summation what would be the last things to get this PR in?

File: src/core/public/ui_settings/ui_settings_client.ts [hunk 5]
```diff
@@ -175,6 +177,17 @@ You can use \`IUiSettingsClient.get("${key}", defaultValue)\`, which will just r
     return this.updateErrors$.asObservable();
   }
 
+  resetThemeVersion() {
+    const key = 'theme:version';
+    // Reset userValue to default value if any legacy theme verion exists.
```
Comment (tmarkley) [inline src/core/public/ui_settings/ui_settings_client.ts:182]: minor: typo `verion` should be `version`

File: src/core/public/ui_settings/ui_settings_client.ts [hunk 6]
```diff
@@ -175,6 +177,17 @@ You can use \`IUiSettingsClient.get("${key}", defaultValue)\`, which will just r
...
+    if (
+      this.cache[key] !== undefined &&
```
Comment (tmarkley) [inline src/core/public/ui_settings/ui_settings_client.ts:184]: Shouldn't we set the value if the cache is undefined?

**Review by tmarkley — CHANGES_REQUESTED** [review]

**Comment by tmarkley** [comment]
> > I think before we merge this we should probably have integration and functional tests that confirm that no matter what setting was applied before updating, `v1` is always applied.
> 
> I think the unit test might be sufficient because it covered the main concern for naming change. Just to make sure we are on same page, `v1` isn't something new but just a renamed theme of `v7`. Compared with previous, only one theme will be loaded no matter what name it is. The only concern would be what is name will displayed to customer. The unit test will verify the override . Functional testing is something we could do but since we actually disabled/hided the from the `Advanced Setting`, not very sure if the value can be retrieved from UI level verification. I don't see much benefit for this since it is covered in unit test

Apologies for not responding to this sooner. The unit test covers the functionality of `resetThemeVersion`, but we don't have anything that confirms whether that function is called in all scenarios and properly sets the theme for all users.

**Comment by kavilla** [comment]
There's a lot of conflicts here and I think the PR still as to address changes. Should we remove the 2.0.0 label?

**Comment by kavilla** [comment]
Will close and  link to the issue we did a temp work around for 2.0 but will need to remove this for 3.0

**Pull request closed without merging**